doc/decisions/header_file_structure.md

```diff
@@ -12,31 +12,128 @@ The big problem is kdbprivate.h. It has two main problems:
-3.  Then there are things that I would call internal or pseudo-private. This includes the struct \_Key and struct \_KeySet, but also keyClearSync or elektraMalloc and the high-level functions needed for codegen. These are things that shouldn't be part of the public API, because we don't want to make all the guarantees associated with that, but nonetheless cannot truly be private. In other words, each of these functions/structs/symbols has a specific "target audience" that may use it.
-    Here I would either create a separate kdbinternal.h or just keep kdbprivate.h, as long as it is a separate file from 4.
-    Symbols belonging to this category should only appear in the private section of the symbols.map file.
-4.  Finally, there is what I'd call unstable or pseudo-public. This includes e.g. ksFindHierarchy or elektraReadArrayNumber. These functions could be public, but for various reasons are not. Maybe they are not well-tested or maybe we just don't want to commit to the function yet. Many of these (but not all) can't be part of libelektra-ease, because the rely on something more private (e.g. directly access the array in a KeySet).
-    Here I would either create a separate kdbinternal.h or just keep kdbprivate.h, as long as it is a separate file from 3.
-    Symbols belonging to this category should only appear in the private section of the symbols.map file.
+### Libraries
+
+A library can be monolithic or modularized.
+Monolithic libraries should be small and bigger libraries with large APIs should be modularized.
+
+A monolithic library `foo` may have these headers (covering categories 3 & 4 from above):
+
+- `src/include/elektra/foo.h`:
+  Contains the full public API of `libelektra-foo`.
+  Will be installed as `<include-root>/elektra/foo.h`.
+- `src/include/elektra/foo/internal.h`:
+  Contains the internal (i.e., non-stable) API of `libelektra-foo`.
+  Will be installed as `<include-root>/elektra/foo/internal.h`.
+
+A modularized library `bar` may have these headers (covering categories 3 & 4 from above):
+
+- `src/include/elektra/foo/public.h`:
+  Declares the public API of `libelektra-foo`, by including `#include <elektra/foo/*.h>`.
+  Will be installed as `<include-root>/elektra/public/foo.h`.
+
+  Such a header may only include `#include <elektra/foo/*.h>` lines.
+
+- `src/include/elektra/foo/*.h`:
+  Additional public API header of `libelektra-foo`.
+  Will be installed as `<include-root>/elektra/foo/*.h`.
+
+  All of these installed headers `installed.h` must be included from `src/lib/foo/public.h` via a line `#include <elektra/foo/installed.h>`.
+
+- `src/include/elektra/foo/internal.h`:
+  Declares the internal (i.e., non-stable) API of `libelektra-foo`, by including `#include <elektra/foo/*.h>`.
+  Will be installed as `<include-root>/elektra/internal/foo.h`.
```

Comment:
Again, it is okay to have internal header files (to be used within other modules of Elektra) but they should not be installed.